### PR TITLE
Improve shared_ptr usage

### PR DESCRIPTION
**Pull request by eric-jj**
The PR will resolve two issues.
1. Use the make_shared to replace std::shared_ptr(new XXX), make_shared will combine the reference count and object create(new) to a memory allocation, but the std::shared_ptr(new XXX) will cause two allocations.
2. Pass the shared_ptr by reference in some cases instead of pass by value. Copy of a shared_ptr is not cheap, which will call the interlock increase and decrease during the function call. Interlock operation will lock the data bus which means all the cores will be paused to wait for it. In my previous project, when we disable the perf counters(which use the interlock operation internal), we almost get 40% latency improvement.

### CODE REVIEW

File: src/ray/common/client_connection.h [hunk 1]
```diff
@@ -62,7 +62,7 @@ template <typename T>
 class ClientConnection;
 
 template <typename T>
-using ClientHandler = std::function<void(std::shared_ptr<ClientConnection<T>>)>;
+using ClientHandler = std::function<void(std::shared_ptr<ClientConnection<T>>&)>;
```
Comment (pcmoritz) [inline src/ray/common/client_connection.h]: Is it possible to make this ```const std::shared_ptr<ClientConnection<T>>&```?

File: src/ray/raylet/worker_pool_test.cc [hunk 2]
```diff
@@ -30,8 +30,8 @@ class WorkerPoolTest : public ::testing::Test {
   WorkerPoolTest() : worker_pool_({}), io_service_() {}
 
   std::shared_ptr<Worker> CreateWorker(pid_t pid) {
-    std::function<void(std::shared_ptr<LocalClientConnection>)> client_handler = [this](
-        std::shared_ptr<LocalClientConnection> client) { HandleNewClient(client); };
+    std::function<void(std::shared_ptr<LocalClientConnection>&)> client_handler = [this](
+        std::shared_ptr<LocalClientConnection>& client) { HandleNewClient(client); };
```
Comment (pcmoritz) [inline src/ray/raylet/worker_pool_test.cc]: try to make these const references

File: src/ray/object_manager/object_manager.h [hunk 3]
```diff
@@ -241,23 +241,23 @@ class ObjectManager {
...
   /// Executes on send_service_ thread pool.
   ray::Status SendObjectData(const ObjectID &object_id,
                              const ObjectBufferPool::ChunkInfo &chunk_info,
-                             std::shared_ptr<SenderConnection> conn);
+                             std::shared_ptr<SenderConnection>& conn);
 
   /// Invoked when a remote object manager pushes an object to this object manager.
   /// This will invoke the object receive on the receive_service_ thread pool.
-  void ReceivePushRequest(std::shared_ptr<TcpClientConnection> conn,
+  void ReceivePushRequest(std::shared_ptr<TcpClientConnection> &conn,
```
Comment (pcmoritz) [inline src/ray/object_manager/object_manager.h]: try to make this a const reference

File: src/ray/object_manager/object_manager.h [hunk 4]
```diff
@@ -241,23 +241,23 @@ class ObjectManager {
...
                                 uint64_t metadata_size, uint64_t chunk_index,
-                                std::shared_ptr<SenderConnection> conn);
+                                std::shared_ptr<SenderConnection>& conn);
 
   /// This method initiates the actual object transfer.
   /// Executes on send_service_ thread pool.
   ray::Status SendObjectData(const ObjectID &object_id,
                              const ObjectBufferPool::ChunkInfo &chunk_info,
-                             std::shared_ptr<SenderConnection> conn);
+                             std::shared_ptr<SenderConnection>& conn);
```
Comment (pcmoritz) [inline src/ray/object_manager/object_manager.h]: try to make this a const reference

File: src/ray/object_manager/object_manager.h [hunk 5]
```diff
@@ -241,23 +241,23 @@ class ObjectManager {
   /// Executes on send_service_ thread pool.
   ray::Status SendObjectHeaders(const ObjectID &object_id, uint64_t data_size,
                                 uint64_t metadata_size, uint64_t chunk_index,
-                                std::shared_ptr<SenderConnection> conn);
+                                std::shared_ptr<SenderConnection>& conn);
```
Comment (pcmoritz) [inline src/ray/object_manager/object_manager.h]: try to make this a const reference

File: src/ray/object_manager/object_manager.h [hunk 6]
```diff
@@ -226,7 +226,7 @@ class ObjectManager {
   /// Synchronously send a pull request via remote object manager connection.
   /// Executes on main_service_ thread.
   ray::Status PullSendRequest(const ObjectID &object_id,
-                              std::shared_ptr<SenderConnection> conn);
+                              std::shared_ptr<SenderConnection>& conn);
```
Comment (pcmoritz) [inline src/ray/object_manager/object_manager.h]: try to make this a const reference

File: src/ray/object_manager/object_manager.h [hunk 7]
```diff
@@ -119,7 +119,7 @@ class ObjectManager {
   /// \param conn The connection.
   /// \param message_type The message type.
   /// \param message A pointer set to the beginning of the message.
-  void ProcessClientMessage(std::shared_ptr<TcpClientConnection> conn,
+  void ProcessClientMessage(std::shared_ptr<TcpClientConnection> &conn,
```
Comment (pcmoritz) [inline src/ray/object_manager/object_manager.h]: try to make this a const reference

**Review by pcmoritz — COMMENTED** [review]
Nice catches! There are also a bunch of linting errors that need to be fixed before we can merge this.

**Comment by eric-jj** [comment]
I found a bug, will reopen it after fix it.

**Comment by robertnishihara** [comment]
It's also fine to leave it open while you work on it. Closing it is fine of course if you prefer that.

**Pull request closed without merging**